WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsViewModel.kt

```diff
@@ -0,0 +1,252 @@
+                items = emptyList()
+        )
+    }
+
+    private fun updateUiStateToFullScreenError() {
+        _uiState.value = VerticalsUiState(
+                showFullscreenError = true,
+                showFullscreenProgress = false,
+                showContent = false,
+                showSkipButton = false,
+                items = emptyList()
+        )
+    }
+
+    private fun updateUiStateToContent(query: String, state: ListState<VerticalModel>) {
+        listState = state
+        _uiState.value = VerticalsUiState(
+                showFullscreenError = false,
+                showFullscreenProgress = false,
+                showContent = true,
+                showSkipButton = StringUtils.isEmpty(query),
+                headerUiState = createHeaderUiState(shouldShowHeader(query), headerInfo),
+                searchInputState = createSearchInputUiState(
+                        query,
+                        showProgress = state is Loading,
+                        hint = headerInfo.inputHint
+                ),
+                items = createSuggestionsUiStates(
+                        onRetry = { updateQuery(query) },
+                        data = state.data,
+                        errorFetchingSuggestions = state is Error
+                )
+        )
+    }
+
+    private fun createSuggestionsUiStates(
+        onRetry: () -> Unit,
+        data: List<VerticalModel>,
+        errorFetchingSuggestions: Boolean
+    ): List<VerticalsListItemUiState> {
+        val items: ArrayList<VerticalsListItemUiState> = ArrayList()
+        if (errorFetchingSuggestions) {
+            val errorUiState = VerticalsFetchSuggestionsErrorUiState(
+                    messageResId = R.string.site_creation_fetch_suggestions_failed,
+                    retryButonResId = R.string.button_retry
+            )
+            errorUiState.onItemTapped = onRetry
+            items.add(errorUiState)
+        } else {
+            val lastItemIndex = data.size - 1
+            data.forEachIndexed { index, model ->
+                items.add(VerticalsModelUiState(model.verticalId, model.name, showDivider = index != lastItemIndex))
+            }
+        }
+        return items
+    }
+
+    private fun shouldShowHeader(query: String): Boolean {
+        return StringUtils.isEmpty(query)
+    }
+
+    private fun createHeaderUiState(
+        isVisible: Boolean,
+        headerInfo: DummyVerticalsHeaderInfoModel
+    ): VerticalsHeaderUiState {
+        return VerticalsHeaderUiState(isVisible, headerInfo.title, headerInfo.subtitle)
+    }
+
+    private fun createSearchInputUiState(
+        query: String,
+        showProgress: Boolean,
+        hint: String
+    ): VerticalsSearchInputUiState {
+        return VerticalsSearchInputUiState(
+                hint,
+                showProgress,
+                showClearButton = !StringUtils.isEmpty(query)
+        )
+    }
+
+    data class VerticalsUiState(
```

Comment:
If I understand it correctly only one of these must be true at any given time: `showFullscreenError`, `showFullscreenProgress`, `showContent`. Wouldn't it be better if we used an enum for it? Better yet, some of these only makes sense for certain `states` (error, progress, content), so maybe we can even go for a sealed class here.

Reply:
I was about to leave a comment for `VerticalsListItemUiState` because I thought in `createSuggestionsUiStates` we were creating multiple `VerticalsFetchSuggestionsErrorUiState` objects. This turned out to be incorrect, but I think my previous suggestion would have helped with that. If it's a success state, we can have multiple items, if it's error, we can only have one and that makes the distinction more clear.

Hope that makes sense!

Reply:
>If I understand it correctly only one of these must be true at any given time: showFullscreenError, showFullscreenProgress, showContent. Wouldn't it be better if we used an enum for it? Better yet, some of these only makes sense for certain states (error, progress, content), so maybe we can even go for a sealed class here.

I think the View should get a complete state of the UI and should need to handle as minimum logic as possible. We could do something like this.
```
  sealed class UiState {
        open val showFullscreenError: Boolean = false
        open val showFullscreenProgress: Boolean = false
        open val showContent: Boolean = false
        open val showSkipButton: Boolean = false
        open val headerUiState: VerticalsHeaderUiState = VerticalsHeaderUiState.Hidden
        open val searchInputState: VerticalsSearchInputUiState = VerticalsSearchInputUiState.Hidden
        open val items: List<VerticalsListItemUiState> = emptyList()

        object FullscreenErrorUiState : UiState() {
            override val showFullscreenError: Boolean = true
        }

        object FullscreenProgressUiState : UiState() {
            override val showFullscreenProgress: Boolean = true
        }

        data class ContentUiState(
            override val showSkipButton: Boolean,
            override val headerUiState: VerticalsHeaderUiState,
            override val searchInputState: VerticalsSearchInputUiState,
            override val items: List<VerticalsListItemUiState>,
            override val showContent: Boolean = true
        ) : UiState()
    }

    sealed class VerticalsSearchInputUiState(val isVisible: Boolean) {
        open val hint: String = ""
        open val showProgress: Boolean = false
        open val showClearButton: Boolean = false

        object Hidden : VerticalsSearchInputUiState(false)
        data class Visible(
            override val hint: String,
            override val showProgress: Boolean,
            override val showClearButton: Boolean
        ) : VerticalsSearchInputUiState(true)
    }

    sealed class VerticalsHeaderUiState(val isVisible: Boolean) {
        open val title: String = ""
        open val subtitle: String = ""

        object Hidden : VerticalsHeaderUiState(false)
        data class Visible(
            override val title: String,
            override val subtitle: String
        ) : VerticalsHeaderUiState(true)
    }
```

If you have a better proposal, I'll be happy to modify it.

>I was about to leave a comment for VerticalsListItemUiState because I thought in createSuggestionsUiStates we were creating multiple VerticalsFetchSuggestionsErrorUiState objects. This turned out to be incorrect, but I think my previous suggestion would have helped with that. If it's a success state, we can have multiple items, if it's error, we can only have one and that makes the distinction more clear.
Hope that makes sense!

If I understand it correctly this would bring a logic to the Fragment. Moreover there is still a chance we'll show "Create a new category/Use what the user typed" when there is an error.